### PR TITLE
Handle font_data in genericjson parser

### PR DESCRIPTION
**Pull request by Funkenjaeger**
Adds font support, fixes arc bug (ref #337)

### CODE REVIEW

**Comment by Funkenjaeger** [comment]
I think this is more or less complete, though I haven't added any changes to address potential surprises associated with kicad-specific text handling.

**Comment by Funkenjaeger** [comment]
Now bypasses kicad-specific overbar handling in render.js.  Also strips control characters from text, which avoids kicad-specific tab-to-space logic and mitigates risk of other possible unforeseen surprises which control characters might cause downstream in javascript.

File: InteractiveHtmlBom/ecad/genericjson.py [hunk 1]
```diff
@@ -65,6 +67,48 @@ def _verify(self, pcb):
...
+        for dwg in self._texts():
+            if 'svgpath' not in dwg:
+                dwg['text'] = self._remove_control_codes(dwg['text'])
+                font_parser.parse_font_for_string(dwg['text'])
+
+        if font_parser.get_parsed_font().items():
+            self.pcbdata['font_data'] = {}
+            for (c, wl) in font_parser.get_parsed_font().items():
+                if c not in self.pcbdata['font_data']:
+                    self.pcbdata['font_data'][c] = wl
```
Comment (qu1ck) [inline InteractiveHtmlBom/ecad/genericjson.py:93]: You can assign get_parsed_font() result directly it's a dict already, no need to deduplicate.

File: InteractiveHtmlBom/ecad/genericjson.py [hunk 2]
```diff
@@ -65,6 +67,48 @@ def _verify(self, pcb):
...
+
+    def _check_font_data(self):
+        self.mc = []
+        for dwg in self._texts():
+            dwg['text'] = self._remove_control_codes(dwg['text'])
+            self.mc += [c for c in dwg['text'] if 'svgpath' not in dwg and
+                        c not in self.pcbdata['font_data']]
+
+        self.mc = list(set(self.mc))  # get unique characters
+        s = ''.join(self.mc)
```
Comment (qu1ck) [inline InteractiveHtmlBom/ecad/genericjson.py:103]: This is hard to follow. Instead of accumulating a large array declare a set, keep adding chars to it (it will ignore elements it already contains) and in the end check if it's empty.

File: InteractiveHtmlBom/ecad/genericjson.py [hunk 3]
```diff
@@ -79,22 +123,29 @@ def _parse(self):
...
 
-        self.logger.info('Successfully parsed {}'.format(self.file_name))
+        if 'font_data' in self.pcbdata:
+            if not self._check_font_data():
+                raise ParsingException(f'Failed parsing {self.file_name}')
+        else:
+            self._parse_font_data()
+            if 'font_data' in self.pcbdata:
+                self.logger.info('No font_data provided in JSON, using '
+                                 'built-in font parser')
```
Comment (qu1ck) [inline InteractiveHtmlBom/ecad/genericjson.py:136]: "using newstroke font" is more informative

File: InteractiveHtmlBom/ecad/genericjson.py [hunk 4]
```diff
@@ -79,22 +123,29 @@ def _parse(self):
                               .format(self.file_name))
             return None, None
 
-        pcbdata = pcb['pcbdata']
-        components = [Component(**c) for c in pcb['components']]
+        self.pcbdata = pcb['pcbdata']
+        self.components = [Component(**c) for c in pcb['components']]
```
Comment (qu1ck) [inline InteractiveHtmlBom/ecad/genericjson.py:127]: Although python allows creating new object fields anywhere it's considered good practice to always declare all fields in constructor, some linters will even enforce it.

But there is no reason for this here, you can just pass pcbdata into functions.

File: InteractiveHtmlBom/ecad/schema/genericjsonpcbdata_v1.schema [hunk 5]
```diff
@@ -602,6 +605,22 @@
             "properties": {
             },
             "title": "ExtraData"
+        },
+        "FontData": {
+            "type": "object",
+            "additionalProperties": {
```
Comment (qu1ck) [inline InteractiveHtmlBom/ecad/schema/genericjsonpcbdata_v1.schema:611]: Use patternProperties with `^?$` regex to enforce single char keys and "additionalProperties": false.

File: InteractiveHtmlBom/ecad/schema/genericjsonpcbdata_v1.schema [hunk 6]
```diff
@@ -602,6 +605,22 @@
...
+                "type": "object",
+                "properties": {
+                    "w": { "type": "number" },
+                    "l": { "$ref": "#/definitions/Polygons" }
```
Comment (qu1ck) [inline InteractiveHtmlBom/ecad/schema/genericjsonpcbdata_v1.schema:615]: This will confuse people because what these coordinates describe is not polygons, even if data has same format. In this case this is an array of polylines.

Can you try adding a definition PolyLineArray that just references Polygons? I think it should work and will be clearer.

File: InteractiveHtmlBom/web/util.js [hunk 7]
```diff
@@ -520,6 +521,8 @@ function initDefaults() {
   settings.markWhenChecked = readStorage("markWhenChecked") || "";
   populateMarkWhenCheckedOptions();
 
+  settings.kicadTextFormatting = config.kicad_text_formatting
```
Comment (qu1ck) [inline InteractiveHtmlBom/web/util.js:524]: No need to have this in settings since user can't change it. You can refer directly to `config` in `render.js`.

**Review by qu1ck — CHANGES_REQUESTED** [review]

**Comment by Funkenjaeger** [comment]
I addressed your comments and rebased on ae27915f08a3f06ba4dc0a8360a181b96c442a8a.  I'll squash if/when you're good with the changes.

**Comment by Funkenjaeger** [comment]
Done

**Comment by qu1ck** [comment]
Thank you for your contribution.
